tests/unittest/llmapi/apps/test_kimi_serve_extensions.py

```
@@ -0,0 +1,548 @@
+
+def dynamic_system_msg(tools: list, **extra) -> dict:
+    return {"role": "system", "content": "", "tools": tools, **extra}
+
+
+class TestToolChoiceValidation:
+    def test_required_with_tools_accepted(self) -> None:
+        req = make_request(tools=[WEATHER_TOOL], tool_choice="required")
+        assert req.tool_choice == "required"
+
+    def test_required_without_tools_rejected(self) -> None:
+        with pytest.raises(ValidationError, match="tools.*must be set"):
```

Ruff (0.16.1)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)